### PR TITLE
[scroll-anchoring] Weird scrolling effect when scrolling YouTube transcript panel

### PR DESCRIPTION
**Pull request by nmoucht**
#### bb9af3d426703c0dab09e5de39fe435eb6876f7f
<pre>
[scroll-anchoring] Weird scrolling effect when scrolling YouTube transcript panel
<a href="https://bugs.webkit.org/show_bug.cgi?id=267005">https://bugs.webkit.org/show_bug.cgi?id=267005</a>
<a href="https://rdar.apple.com/120378263">rdar://120378263</a>

Reviewed by Simon Fraser.

After <a href="https://commits.webkit.org/272277@main">https://commits.webkit.org/272277@main</a> we descend subscrollers that are not maintaining a
scroll anchor. We need to replicate the logic used in ScrollAnchoringController::examineAnchorCandidate
for ScrollAnchoringController::didFindPriorityCandidate, to ensure we don&apos;t end up choosing an anchor
element that is a descendant of a scrollable area that is maintaining a scroll anchor.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/focused-element-nested-anchor-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/focused-element-nested-anchor.html: Added.
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::scrollAnchoringControllerForElement):
(WebCore::ScrollAnchoringController::didFindPriorityCandidate):
(WebCore::canDescendIntoElement):
(WebCore::ScrollAnchoringController::isExcludedSubtree):
(WebCore::ScrollAnchoringController::examineAnchorCandidate):
(WebCore::ScrollAnchoringController::chooseAnchorElement):
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:

Canonical link: <a href="https://commits.webkit.org/272616@main">https://commits.webkit.org/272616@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/97959b7bbf7c61fb5e5c9fc202ffc136edb43ce9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32278 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34834 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29237 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13365 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28795 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9283 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28847 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8087 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8230 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28773 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-src.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36177 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29346 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34354 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6295 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32218 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10011 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28542 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8994 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4184 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8911 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp [hunk 1]
```diff
@@ -139,6 +139,19 @@ static RefPtr<Element> anchorElementForPriorityCandidate(Element* element)
     return nullptr;
 }
 
+static ScrollAnchoringController* scrollAnchoringControllerForElement(Element& element)
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:142]: It's confusing that we have this and `RenderObject::findScrollAnchoringControllerForRenderer()` with similar names, yet `findScrollAnchoringControllerForRenderer()` walks the ancestor tree, but this does not. They need more distinct names.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp [hunk 2]
```diff
@@ -150,7 +163,8 @@ bool ScrollAnchoringController::didFindPriorityCandidate(Document& document)
 
         while (iterElement && iterElement.get() != elementForScrollableArea(m_owningScrollableArea)) {
             auto candidateResult = examineAnchorCandidate(*iterElement);
-            if (candidateResult == CandidateExaminationResult::Exclude || (iterElement == candidateElement && candidateResult == CandidateExaminationResult::Skip))
+            auto scrollAnchoringController = scrollAnchoringControllerForElement(*iterElement);
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: I'd call this `elementsController` or something.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp [hunk 3]
```diff
@@ -150,7 +163,8 @@ bool ScrollAnchoringController::didFindPriorityCandidate(Document& document)
...
+            if (candidateResult == CandidateExaminationResult::Exclude || (iterElement == candidateElement && candidateResult == CandidateExaminationResult::Skip) || (scrollAnchoringController && scrollAnchoringController->anchorElement()))
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: This logic is getting hard to follow; a lambda might be better. I'm not clear on why `scrollAnchoringController` having its own anchor means we return null here. In general, be careful about "reaching across" into another object; we're in `ScrollAnchoringController` code here, but it's doing something with a different `ScrollAnchoringController` and this gets confusing.

For example, how can I tell if `scrollAnchoringController` is for a descendant of `this`, or a sibling or ancestor? If there's code that needs to reason about multiple ScrollAnchoringControllers, consider putting it into a static function.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp [hunk 4]
```diff
@@ -164,7 +178,10 @@ bool ScrollAnchoringController::didFindPriorityCandidate(Document& document)
     if (RefPtr priorityCandidate = viablePriorityCandidateForElement(document.focusedElement())) {
         m_anchorElement = priorityCandidate;
         m_lastOffsetForAnchorElement = computeOffsetFromOwningScroller(*m_anchorElement->renderer());
-        LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::viablePriorityCandidateForElement() for scroller: " << m_owningScrollableArea << " found priority candidate: " << *priorityCandidate);
+        if (auto element = elementForScrollableArea(m_owningScrollableArea))
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: It's better to not have conditional code that only affects logging (we pay the cost for `elementForScrollableArea()` in release builds).

You can use `ValueOrNull` to log a pointer if necessary.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp [hunk 5]
```diff
@@ -289,7 +298,8 @@ Element* ScrollAnchoringController::findAnchorElementRecursive(Element* element)
 
 void ScrollAnchoringController::chooseAnchorElement(Document& document)
 {
-    LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::chooseAnchorElement() starting findAnchorElementRecursive: ");
+    if (auto element = elementForScrollableArea(m_owningScrollableArea))
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: Ditto.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp (hunk 1 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:142]: Right I think findScrollAnchoringControllerForRenderer is inadequately named. Changed this to searchParentChainForScrollAnchoringController to make it more clear we are walking the ancestor tree.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp (hunk 2 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp (hunk 3 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: Right this is getting a bit confusing. Separated this to a static function and added a few comments.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp (hunk 4 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollAnchoringController.cpp (hunk 5 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollAnchoringController.cpp]: Fixed.
